virtcontainers/sandbox.go

```diff
@@ -967,6 +971,106 @@ func (s *Sandbox) removeNetwork() error {
 }
 
+func (s *Sandbox) generateNetInfo(inf *grpc.Interface) (NetworkInfo, error) {
+	hw, err := net.ParseMAC(inf.HwAddr)
+	if err != nil {
+		return NetworkInfo{}, err
+	}
+
+	var addrs []netlink.Addr
+	for _, addr := range inf.IPAddresses {
+		netlinkAddrStr := fmt.Sprintf("%s/%s", addr.Address, addr.Mask)
+		netlinkAddr, err := netlink.ParseAddr(netlinkAddrStr)
+		if err != nil {
+			return NetworkInfo{}, fmt.Errorf("could not parse %q: %v", netlinkAddrStr, err)
+		}
+
+		addrs = append(addrs, *netlinkAddr)
+	}
+
+	return NetworkInfo{
+		Iface: NetlinkIface{
+			LinkAttrs: netlink.LinkAttrs{
+				Name:         inf.Name,
+				HardwareAddr: hw,
+				MTU:          int(inf.Mtu),
+			},
+			Type: "",
+		},
+		Addrs: addrs,
+	}, nil
+}
+
+// AddInterface adds new nic to the sandbox.
+func (s *Sandbox) AddInterface(inf *grpc.Interface) (*grpc.Interface, error) {
+	netInfo, err := s.generateNetInfo(inf)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint, err := createVirtualNetworkEndpoint(len(s.networkNS.Endpoints), inf.Name, inf.Device, inf.HwAddr, s.config.NetworkConfig.InterworkingModel)
+	if err != nil {
+		return nil, err
+	}
+	endpoint.SetProperties(netInfo)
+	if err := doNetNS(s.networkNS.NetNsPath, func(_ ns.NetNS) error {
+		return endpoint.HotAttach(s.hypervisor)
+	}); err != nil {
+		return nil, err
+	}
+
+	// Update the sandbox storage
+	s.networkNS.Endpoints = append(s.networkNS.Endpoints, endpoint)
+	if err := s.storage.storeSandboxNetwork(s.id, s.networkNS); err != nil {
+		return nil, err
+	}
+
+	// Add network for vm
+	return s.agent.updateInterface(inf)
+}
+
+// RemoveInterface removes a nic of the sandbox.
+func (s *Sandbox) RemoveInterface(inf *grpc.Interface) (*grpc.Interface, error) {
+	for i, endpoint := range s.networkNS.Endpoints {
+		if endpoint.HardwareAddr() == inf.HwAddr {
+			if err := endpoint.HotDetach(s.hypervisor, s.networkNS.NetNsCreated, s.networkNS.NetNsPath); err != nil {
+				return inf, err
+			}
+			s.networkNS.Endpoints = append(s.networkNS.Endpoints[:i], s.networkNS.Endpoints[i+1:]...)
+			break
+		}
+	}
+	return nil, nil
+}
+
+// ListInterfaces lists all nics and their configurations in the sandbox.
+func (s *Sandbox) ListInterfaces() ([]*grpc.Interface, error) {
+	return s.agent.listInterfaces()
+}
+
+// UpdateRoutes updates the sandbox route table (e.g. for portmapping support).
+func (s *Sandbox) UpdateRoutes(routes []*grpc.Route) ([]*grpc.Route, error) {
+	return s.agent.updateRoutes(routes)
+}
+
+// ListRoutes lists all routes and their configurations in the sandbox.
+func (s *Sandbox) ListRoutes() ([]*grpc.Route, error) {
+	return s.agent.listRoutes()
+}
+
+// deleteNetworks deletes all existing network interfaces.
+func (s *Sandbox) deleteNetworks() error {
+	for i, endpoint := range s.networkNS.Endpoints {
+		if err := endpoint.HotDetach(s.hypervisor, s.networkNS.NetNsCreated, s.networkNS.NetNsPath); err != nil {
+			s.networkNS.Endpoints = s.networkNS.Endpoints[i:]
```

Comment:
`panic: runtime error: slice bounds out of range`

you can't iterate over `s.networkNS.Endpoints` and update it at the same time

see https://play.golang.org/p/2yj9Ahrzsuu

Reply:
Just update slice and end iteration when an endpoint detaching returns an error not every time. Is this correct?

Reply:
You've got to copy the slice and update that copy to remove all the endpoints that were successfully deleted to avoid the issue @devimc has pointed out.

Reply:
OK. But I think it's safe. In other codes
https://github.com/kata-containers/runtime/blob/fc0142ec8e403b784a7e240b04d4b09b7b552426/virtcontainers/sandbox.go#L1108-L1113

Reply:
Good point. I will update this. In the end, I find that updating s.networkNS.Endpoints seems useless...